portal/flatpak-portal.c

```diff
@@ -700,6 +709,702 @@ authorize_method_handler (GDBusInterfaceSkeleton *interface,
   return TRUE;
 }
 
+static void
+monitor_register (PortalFlatpakUpdateMonitor *monitor,
+                  const char *handle)
+{
+  G_LOCK (monitors);
+  g_hash_table_insert (monitors, g_strdup (handle), g_object_ref (monitor));
+  G_UNLOCK (monitors);
+}
+
+static void
+monitor_unregister (const char *handle)
+{
+  G_LOCK (monitors);
+  g_hash_table_remove (monitors, handle);
+  G_UNLOCK (monitors);
+}
+
+typedef struct {
+  guint timeout;
+  char *sender;
+  char *id;
+  char *name;
+  char *arch;
+  char *branch;
+  char *commit;
+  GKeyFile *app_info;
+  GCancellable *cancellable;
+  gboolean installing;
+} MonitorData;
+
+static void
+monitor_data_free (gpointer data)
+{
+  MonitorData *m = data;
+
+  if (m->timeout)
+    g_source_remove (m->timeout);
+  g_free (m->sender);
+  g_free (m->id);
+  g_free (m->name);
+  g_free (m->arch);
+  g_free (m->branch);
+  g_free (m->commit);
+  g_key_file_unref (m->app_info);
+  if (m->cancellable)
+    g_object_unref (m->cancellable);
+
+  g_free (m);
+}
+
+static MonitorData *
+monitor_get_data (PortalFlatpakUpdateMonitor *monitor)
+{
+  return (MonitorData *)g_object_get_data (G_OBJECT (monitor), "data");
+}
+
+static void
+monitor_close (PortalFlatpakUpdateMonitor *monitor)
+{
+  const char *handle;
+  MonitorData *m;
+
+  handle = g_dbus_interface_skeleton_get_object_path (G_DBUS_INTERFACE_SKELETON (monitor));
+
+  if (handle)
+    g_dbus_interface_skeleton_unexport (G_DBUS_INTERFACE_SKELETON (monitor));
+
+  m = monitor_get_data (monitor);
+
+  if (m->timeout)
+    {
+      g_source_remove (m->timeout);
+      m->timeout = 0;
+    }
+
+  if (m->cancellable)
+    g_cancellable_cancel (m->cancellable);
+
+  monitor_unregister (handle);
+}
+
+static void
+close_monitors_in_thread_func (GTask *task,
+                               gpointer source_object,
+                               gpointer task_data,
+                               GCancellable *cancellable)
+{
+  const char *sender = (const char *)task_data;
+  GSList *list = NULL;
+  GSList *l;
+  GHashTableIter iter;
+  PortalFlatpakUpdateMonitor *monitor;
+
+  G_LOCK (monitors);
+  if (monitors)
+    {
+      g_hash_table_iter_init (&iter, monitors);
+      while (g_hash_table_iter_next (&iter, NULL, (gpointer *)&monitor))
+        {
+          const char *s;
+
+          s = (const char *)g_object_get_data (G_OBJECT (monitor), "sender");
+          if (strcmp (sender, s) == 0)
+            list = g_slist_prepend (list, g_object_ref (monitor));
+        }
+    }
+  G_UNLOCK (monitors);
+
```

Comment:
It seems like a race condition to keep the monitor object in the monitors hashtable and then call close on them below. Whats protecting some other thread to run e.g. andle_close() and double-close the handle.